src/IB/IBImplicitStaggeredHierarchyIntegrator.cpp

```diff
@@ -606,12 +894,41 @@ PetscErrorCode IBImplicitStaggeredHierarchyIntegrator::compositeIBJacobianSetup(
     IBTK_CHKERRQ(ierr);
     ierr = MatAssemblyEnd(*A, MAT_FINAL_ASSEMBLY);
     IBTK_CHKERRQ(ierr);
-    Vec* component_sol_vecs;
-    ierr = VecMultiVecGetSubVecs(x, &component_sol_vecs);
-    IBTK_CHKERRQ(ierr);
-    Vec X = component_sol_vecs[1];
-    d_ib_implicit_ops->setLinearizedPosition(X);
-    return 0;
+    if (d_solve_for_position)
+    {
+        Vec* component_sol_vecs;
+        ierr = VecMultiVecGetSubVecs(x, &component_sol_vecs);
+        IBTK_CHKERRQ(ierr);
+        Vec X = component_sol_vecs[1];
+        d_ib_implicit_ops->setLinearizedPosition(X);
+    }
+    else
+    {
+        // Get the estimate of X^{n+1} from the current iterate U^{n+1} and set as it as
+        // a base vector in matrix-free Lagrangian force Jacobian.
+        const double half_time = d_integrator_time + 0.5 * d_current_dt;
+        Vec X_new;
+        ierr = VecDuplicate(d_X_current, &X_new);
+        IBTK_CHKERRQ(ierr);
+
+        Pointer<SAMRAIVectorReal<NDIM, double> > u = PETScSAMRAIVectorReal::getSAMRAIVector(x);
+        VariableDatabase<NDIM>* var_db = VariableDatabase<NDIM>::getDatabase();
+        Pointer<VariableContext> current_ctx = d_ins_hier_integrator->getCurrentContext();
+        Pointer<Variable<NDIM> > u_var = d_ins_hier_integrator->getVelocityVariable();
+        const int u_current_idx = var_db->mapVariableAndContextToIndex(u_var, current_ctx);
+        const int u_new_idx = u->getComponentDescriptorIndex(0);
+        d_hier_velocity_data_ops->linearSum(d_u_idx, -0.5, u_new_idx, -0.5, u_current_idx);
+
+        d_ib_implicit_ops->interpolateLinearizedVelocity(
+            d_u_idx, getCoarsenSchedules(d_object_name + "::u::CONSERVATIVE_COARSEN"),
+            getGhostfillRefineSchedules(d_object_name + "::u"), half_time);
+        d_ib_implicit_ops->computeLinearizedResidual(d_X_current, X_new);
+        d_ib_implicit_ops->setLinearizedPosition(X_new);
+
+        ierr = VecDestroy(&X_new);
+        IBTK_CHKERRQ(ierr);
+    }
+    return ierr;
 } // compositeIBJacobianSetup
 
 PetscErrorCode IBImplicitStaggeredHierarchyIntegrator::compositeIBJacobianApply_SAMRAI(Mat A, Vec x, Vec f)
```

Comment:
I thought about it, but then I wanted to avoid having so many different static functions in this class. There are a lot already. One possibility is to remove keyword composite from them. But I am also OK with having different functions like  {IBJacobianSetup_velocity, IBJacobianSetup_position} if that makes it cleaner?